### PR TITLE
ISPN-7754 forEach method

### PR DESCRIPTION
**Pull request by karesti**
https://issues.jboss.org/browse/ISPN-7754

### CODE REVIEW

File: core/src/main/java/org/infinispan/Cache.java [hunk 1]
```diff
@@ -443,4 +444,18 @@ default V compute(K key,
...
    }
+
+   /**
+    * {@inheritDoc}
+    * <p>
+    * When this method is used on a clustered cache, either replicated or distributed, the biconsumer will be serialized
+    * to owning nodes to perform the operation in the most performant way. However this means the biconsumer must
+    * have an appropriate {@link org.infinispan.commons.marshall.Externalizer} or be {@link java.io.Serializable} itself.
+    * <p>
+    * For transactional caches, whenever the values of the caches are collections, and the mapping function modifies the collection, the collection
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/Cache.java]: It's not specific to collections: the values in cache must not be mutated, period, and we'll rather enforce that than remove the limitation.

File: core/src/main/java/org/infinispan/Cache.java [hunk 2]
```diff
@@ -443,4 +444,18 @@ default V compute(K key,
...
+    * must be copied and not directly modified, otherwise whenever rollback is called it won't work.
+    * This limitation could disappear in following releases if technically possible.
+    */
+   @Override
+   void forEach(BiConsumer<? super K, ? super V> action);
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/Cache.java]: Where is the `SerializableBiConsumer` overload?

File: core/src/main/java/org/infinispan/security/impl/SecureCacheImpl.java [hunk 3]
```diff
@@ -357,6 +358,12 @@ public V compute(K key, BiFunction<? super K, ? super V, ? extends V> remappingF
    }
 
    @Override
+   public void forEach(BiConsumer<? super K, ? super V> action) {
+      authzManager.checkPermission(subject, AuthorizationPermission.WRITE);
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/security/impl/SecureCacheImpl.java]: Shouldn't this be READ?

File: core/src/main/java/org/infinispan/Cache.java (hunk 1 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/Cache.java]: Err, actually this is a consumer, and you should not modify the collection - what would you do with the copy?

File: core/src/test/java/org/infinispan/distribution/DistSyncFuncTest.java [hunk 4]
```diff
@@ -396,4 +400,38 @@ public void testComputeIfAbsentFromNonOwner() throws InterruptedException {
...
+   static final Map<Object, String> forEachStructure = new ConcurrentHashMap<>();
+   static final AtomicInteger forEachOffset = new AtomicInteger();
+
+   static int populateNextForEachStructure(String obj) {
+      int offset = forEachOffset.getAndIncrement();
+      forEachStructure.put(offset, obj);
+      return offset;
+   }
+
+   public void testForEach() throws Exception {
```
Comment (rvansa) [inline core/src/test/java/org/infinispan/distribution/DistSyncFuncTest.java]: This test is silly. Accessing a static variable on (potentially) remote node...

File: core/src/main/java/org/infinispan/security/impl/SecureCacheImpl.java (hunk 3 above)
Comment (karesti) [inline core/src/main/java/org/infinispan/security/impl/SecureCacheImpl.java]: yes, it should be

File: core/src/main/java/org/infinispan/Cache.java (hunk 1 above)
Comment (karesti) [inline core/src/main/java/org/infinispan/Cache.java]: this is me that does copy/paste without thinking twice sometimes. So well, the comment is not correct, indeed !!

File: core/src/test/java/org/infinispan/distribution/DistSyncFuncTest.java (hunk 4 above)
Comment (karesti) [inline core/src/test/java/org/infinispan/distribution/DistSyncFuncTest.java]: 😢

**Comment by tristantarrant** [comment]
Temporarily closing because of CI overload

**Pull request closed without merging**